platforms/software/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/results/SerializableTestResultStore.java

```diff
@@ -0,0 +1,363 @@
+
+        @Override
+        public void metadata(TestDescriptorInternal testDescriptor, TestMetadataEvent event) {
+            // Ignored for now.
+        }
+
+        @Override
+        public void close() throws IOException {
+            // Write a 0 id to terminate the file
+            resultsEncoder.writeSmallLong(0);
+            CompositeStoppable.stoppable(resultsEncoder, outputZipFileSystem).stop();
+        }
+    }
+
+    public boolean hasResults() {
+        if (Files.exists(serializedResultsFile) && Files.exists(outputZipFile)) {
+            // Inspect the results file, read first ID to see if there are any results
+            try (KryoBackedDecoder decoder = openAndInitializeDecoder()) {
+                return decoder.readSmallLong() != 0;
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        } else {
+            return false;
+        }
+    }
+
+    @NonNullApi
+    public static final class StoredResult {
+        private final long id;
+        private final SerializableTestResult testResult;
+        private final long parentId;
+
+        private StoredResult(long id, SerializableTestResult testResult, long parentId) {
+            this.id = id;
+            this.testResult = testResult;
+            this.parentId = parentId;
+        }
+
+        public SerializableTestResult getTestResult() {
+            return testResult;
+        }
+
+        public Object getId() {
+            return id;
+        }
+
+        @Nullable
+        public Object getParentId() {
+            return parentId == 0 ? null : parentId;
+        }
+    }
+
+    /**
+     * Get the root result of the test results.
+     *
+     * @return the root result
+     * @throws IllegalStateException if there are no results
+     */
+    public StoredResult getRootResult() throws IOException {
+        if (!hasResults()) {
+            throw new IllegalStateException("No results");
+        }
+        StoredResult root = findInResults(result -> {
+            if (result.getParentId() == null) {
+                return result;
+            }
+            return null;
+        });
+        if (root == null) {
+            throw new IllegalStateException("No root result");
+        }
+        return root;
+    }
+
+    /**
+     * Visit every result in the store. Parents are visited <em>AFTER</em> their children, but not necessarily in a breadth-first or depth-first order.
+     * The action is called once for each result.
+     *
+     * @param action the action to perform on each result
+     * @throws IOException if an error occurs while reading the results
+     */
+    public void forEachResult(Consumer<? super StoredResult> action) throws IOException {
+        findInResults(result -> {
+            action.accept(result);
+            return null;
+        });
+    }
+
+    @Nullable
+    private <T> T findInResults(Function<? super StoredResult, T> action) throws IOException {
```

Comment:
💭 If you make this return `Optional<T>` it gets rid of the need for null checks and allows more concise use of `.orElseThrow()` from the caller.